File: src/test/baekjoon/python/stack/ValueInParentheses.py
parentheses = list(input())
stack = []
answer = 0
temp_num = 1

for i in range(len(parentheses)):
    p = parentheses[i]
    if p == '(':
        stack.append(p)
        temp_num *= 2
    if p == '[':
        stack.append(p)
        temp_num *= 3
    if p == ')':
        if not stack or stack[-1] == '[':
            answer = 0
            break
        if parentheses[i - 1] == '(':
            answer += temp_num
        stack.pop()
        temp_num //= 2
    if p == ']':
        if not stack or stack[-1] == '(':
            answer = 0
            break
        if parentheses[i - 1] == '[':
            answer += temp_num
        stack.pop()
        temp_num //= 3

if stack:
    print(0)
else:
    print(answer)

File: src/test/baekjoon/java/stack/ValueInParentheses.java
package test.baekjoon.java.stack;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class ValueInParentheses {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] parentheses = br.readLine().split("");
        Stack<String> stack = new Stack<>();
        int answer = 0;
        int tempNum = 1;

        for (int i = 0; i < parentheses.length; i++) {
            String p = parentheses[i];

            if (p.equals("(")) {
                stack.push(p);
                tempNum *= 2;
            }

            if (p.equals("[")) {
                stack.push(p);
                tempNum *= 3;
            }

            if (p.equals(")")) {
                if (stack.isEmpty() || stack.peek().equals("[")) {
                    answer = 0;
                    break;
                }

                if (parentheses[i - 1].equals("(")) {
                    answer += tempNum;
                }

                stack.pop();
                tempNum /= 2;
            }

            if (p.equals("]")) {
                if (stack.isEmpty() || stack.peek().equals("(")) {
                    answer = 0;
                    break;
                }

                if (parentheses[i - 1].equals("[")) {
                    answer += tempNum;
                }

                stack.pop();
                tempNum /= 3;
            }
        }

        if (stack.isEmpty()) {
            System.out.println(answer);
        } else {
            System.out.println(0);
        }
    }
}
